Background image uploads: product creation case

hichamboushaba opened the pull request:
Closes: #4742 and #4743

### Description
This PR updates logic of image uploads to handle products under creation, this is a summary of the flows:

1. If the update has been assigned an id (user already clicked on `save as draft` or `publish`), then the product will follow the same behavior as editing a regular product.
2. If the user saves the product while there are ongoing uploads, we'll cancel them and recreate new ones using the new product id (as the products are sequential, it means we'll cancel only the ongoing one, and remove the other ones from the queue).
3. If the user tries to leave the product screen while there are still ongoing uploads, and the product hasn't been saved, then we will do the following:
   - First, we will stop observing upload changes, this will make the handler responsible for the uploads, so it will cache the successful events.
   - If the user decides to discard changes, then we will just cancel uploads.
   - If the user decides to save the product as a draft, then the same flow as saving the product will take place, which means that the uploads will be assigned to the new product id, and be uploaded/saved in the background.
   - If the user decides to return to the product, then we will restart observing image upload changes.

### Testing instructions
The test cases will match the above changes.

##### Case 1
1. Click on create a product.
2. Make some changes and save the product (or publish it).
3. Add some images to the product (from local device)
4. While the images are still being uploaded, leave the product screen.
5. Confirm that the snackbar about image uploads is displayed, and that they continue the upload in the background.
6. Confirm that the product is updated afterwards with the uploaded images.

##### Case 2
1. Click on create a product.
2. Add some images to the product (from local device)
3. While the images are still being uploaded, click on save as draft (or publish).
4. Confirm that the ongoing and successful uploads are assigned to the new product.
5. Confirm that the upload continues, and that you can both wait and save the product, or leave the uploads to continue in the background.

##### Case 3
1. Click on create a product.
2. Add some images to the product (from local device)
3. While the images are still being uploaded, leave the product screen.
4. Confirm that the discard changes dialog is displayed.
5. Click on Save as draft button.
6. Confirm that once the product is saved, the upload continues in the background.
7. Confirm that the product is successfully updated with the new images.

##### Case 4
1. Click on create a product.
2. Add some images to the product (from local device)
3. While the images are still being uploaded, Swipe the app away from the recent app menu
4. Confirm that the upload is cancelled.

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

Commit:
Add a function to cancel and reschedule ongoing uploads for new products

Commit:
Re-assign uploads to the new remote id when a product is created

Commit:
Update productId of past successful uploads for created products

Commit:
Handle case of saving draft product from the dialog

Commit:
Re-introduce the images discard dialog if the product hasn't been saved

Commit:
Fix formatting

Commit:
Restart observing image uploads only when not exiting

Commit:
Add test cases for product creation flow

Commit:
Update the logic of canceling uploads for creating products

Commit:
Add test cases for handling assigning uploads to created product

Commit:
Make sure to set postId before uploading files

Commit:
Allow saving a non created product with ongoing uploads

Commit:
Update tests

Commit:
Update condition to make it more easier to understand

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/media/ProductImagesUploadWorker.kt:
```diff
@@ -199,6 +199,7 @@ class ProductImagesUploadWorker @Inject constructor(
             val doneUploads = uploadList.count { it.isDone }
             notificationHandler.update(doneUploads + 1, uploadList.size)
             try {
+                work.fetchedMedia.postId = work.productId
```
This is not related to this PR, but it's something I missed in the previous PR #4720

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt:
```diff
@@ -533,43 +537,42 @@ class ProductDetailViewModel @Inject constructor(
             triggerEvent(
                 ShowDialog(
+                    messageId = message,
+                    positiveButtonId = string.discard,
+                    negativeButtonId = string.keep_editing,
```
I added an explicit to those, since we previously has been depending on default values specified here https://github.com/woocommerce/woocommerce-android/blob/697172b843b1357f402808f24d829fe13f91a721/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dialog/WooDialog.kt#L39-L40 And those can be changed for any reason, and IMO we should remove them.

Commit:
Use the newProductId explicitly

0nko reviewed (approved):
All the test cases work as expected, nice work 👍. 

I just noticed that I'm seeing a memory leak in CanaryLeak in the `ProductDetailFragment` (gallery view). This probably isn't related, though, so we can look at that later.

0nko commented:
We can merge after the conflicts are resolved.

Commit:
Merge remote-tracking branch 'origin/develop' into issue/4742-background-uploads-add-product

# Conflicts:
#	WooCommerce/src/test/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModelTest.kt

Commit:
Move test cases to the correct file

hichamboushaba commented:
Thanks @0nko for the review, I fixed the conflict, and I just noticed that we have a specific file for tests related to the creation flow, I moved the tests to it in 0e40d3c, if you want to take a second look 🙏

0nko commented:
Thanks! There's just some detekt issue.

Commit:
Fix detekt issue
